server/app/services/application/ApplicationEventDetails.java

```diff
@@ -0,0 +1,101 @@
+package services.application;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.auto.value.AutoValue;
+import java.util.Optional;
+
+/**
+ * The details of a single Application event based on its type.
+ *
+ * <p>This class is intended to represent the specific relevant details of some event related to an
+ * {@code Application}. The event is identified by {code EventType} and its corresponding data; as
+ * such only one of the Optional detail fields should be present. This class will be json serialized
+ * into a database column.
+ */
+@AutoValue
+@JsonDeserialize(builder = AutoValue_ApplicationEventDetails.Builder.class)
+public abstract class ApplicationEventDetails {
+  /** The event affecting the application. */
+  @JsonProperty("event_type")
+  public abstract Type eventType();
+
+  // Only one of the following Event fields should be set.
```

Comment:
Is there anything we can build into the class to enforce this constraint?

One thing that might make sense would be to forgo having the AutoValue.Builder below since it gives the illusion that multiple fields could be set. Instead, we could have two static methods:

```
public abstract class ApplicationEventDetails {
...
...

  public static ApplicationEventDetails createStatusEvent(StatusEvent ev) {
    return new ApplicationEventDetails(EventType.STATUS, Optional.of(ev), Optional.empty());
  }
}
```

A downside to this is that as the number of events grows, the number of arguments to the constructor will also grow. However, constraining this complexity to the class itself would help prevent misuse where multiple fields are set.

Reply:
Yes! I forgot about that.  I added builder validation to ensure only one is set.